File: 2023/python/solutions/day04.py
from lib.python import ProblemInput, ProblemResult, Solution


class Solution04(Solution):
    @staticmethod
    def results() -> list[ProblemResult]:
        return [13, 26443, 30, 6284877]

    @staticmethod
    def solve_part01(data: ProblemInput, is_sample: bool) -> ProblemResult:
        return sum(compute_value(winning, mine) for winning, mine in parse(data))

    @staticmethod
    def solve_part02(data: ProblemInput, is_sample: bool) -> ProblemResult:
        copies = [1] * len(data)
        for i, (winning, mine) in enumerate(parse(data)):
            value = count_hits(winning, mine)
            for j in range(i + 1, i + 1 + value):
                copies[j] += copies[i]

        return sum(copies)


def parse(data: ProblemInput) -> list[tuple[set[int], set[int]]]:
    cards = []
    for line in data:
        left_str, right_str = line.split(":")[1].split("|")
        left = {int(n) for n in left_str.strip().split()}
        right = {int(n) for n in right_str.strip().split()}
        cards.append((left, right))
    return cards


def compute_value(winning: set[int], mine: set[int]) -> int:
    h = count_hits(winning, mine)
    return 1 << (h - 1) if h else 0


def count_hits(winning: set[int], mine: set[int]) -> int:
    return len(winning.intersection(mine))

File: 2023/rust/solutions/day04.rs
use aoc_lib::iterator::ParsedExt;
use itertools::Itertools;

use aoc_lib::prelude::solution::Solution;
use aoc_lib::prelude::types::{ProblemInput, ProblemResult, ToResult};
use std::collections::HashSet;

type Hand = HashSet<u8>;

pub struct Solution04;

impl Solution04 {
    fn parse(&self, data: ProblemInput) -> Vec<(Hand, Hand)> {
        let mut cards = vec![];
        for line in data.lines().iter() {
            let hand_strs = line.split(':').nth(1).unwrap().split('|');
            let parsed = hand_strs
                .map(|s| s.split_whitespace().parsed().collect::<HashSet<_>>())
                .take(2)
                .collect_tuple()
                .unwrap();
            cards.push(parsed)
        }

        cards
    }

    fn compute_value(&self, winning: &Hand, mine: &Hand) -> u32 {
        match self.count_hits(winning, mine) {
            0 => 0,
            h => 1 << (h - 1),
        }
    }

    fn count_hits(&self, winning: &Hand, mine: &Hand) -> usize {
        winning.intersection(mine).count()
    }
}

impl Solution for Solution04 {
    fn results(&self) -> [ProblemResult; 4] {
        [
            ProblemResult::Unsolved,
            ProblemResult::Unsolved,
            ProblemResult::Unsolved,
            ProblemResult::Unsolved,
        ]
    }

    fn solve_version01(&self, input: ProblemInput, _is_sample: bool) -> ProblemResult {
        self.parse(input)
            .iter()
            .map(|(winning, mine)| self.compute_value(winning, mine))
            .sum::<u32>()
            .to_result()
    }

    fn solve_version02(&self, input: ProblemInput, _is_sample: bool) -> ProblemResult {
        let mut copies = vec![1u32; input.lines().len()];
        for (i, (winning, mine)) in self.parse(input).iter().enumerate() {
            let value = self.count_hits(winning, mine);
            let current = *copies.get(i).unwrap();
            for j in (i + 1)..=(i + value) {
                *copies.get_mut(j).unwrap() += current;
            }
        }

        copies.iter().sum::<u32>().to_result()
    }
}
